### PR TITLE
#180 | Resolved The User Assignment Existence Bug in Assignment Search Flow

### PR DESCRIPTION
**Pull request by anilhazar**
# PULL REQUEST TEMPLATE

## Description

Implemented User's Assigment's Existence Check in Assignment Search Flow.

## Related Issue

#180 

## Proposed Changes

Please provide a brief and descriptive title for your pull request, summarizing the changes or features introduced. It
should be in English and clearly indicate the purpose of the pull request. Optionally, you can provide a more detailed
description of what you have done.

- Branches should follow the below `Naming Conventions of Branchs` header in [README.md](https://github.com/afet-yonetim-sistemi/ays-be/edit/main/README.md) File

## Additional Information

If there's any additional information or context that the reviewers should know, please provide it here.

## Checklist

Please check the following before submitting your pull request:

- [X] I have tested my changes locally and they are working as expected.
- [X] The code is formatted according to the project's coding guidelines and style.
- [X] Necessary documentation has been added or existing documentation has been updated.
- [X] Relevant unit tests have been written and included.
- [X] Relevant integration tests have been written and included.
- [X] The code has been reviewed to ensure its quality.
- [X] The code does not contain any issues flagged by SonarLint.
- [X] Default reviewers have been assigned to this pull request.
- [X] Assignees have been added if necessary.
- [X] Labels have been applied if necessary.
- [X] If an issue is opened and this pull request is related to that issue, it should not be associated with any project
  board or milestone. However, if there is no issue, the pull request should be associated with the appropriate project
  board (BE Board, In Review column) and milestone to track the progress.

## Reviewers

The default project maintainers and team members will be automatically assigned as reviewers for this pull request.

### CODE REVIEW

File: src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    @Serial
+    private static final long serialVersionUID = -3240549605152804148L;
+
+    /**
+     * Constructs a new AysAssignmentUserAlreadyAssigned with the specified userId.
+     *
+     * @param userId the userId of user that will take assignment.
+     */
+    public AysAssignmentUserAlreadyAssigned(String userId) {
+        super("USER ALREADY HAS AN ASSIGNMENT! userId:" + userId);
```
Comment (moaydogdu) [inline src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java:18]: Statik bir default mesaj halinde tanımlanabilir.

**Review by moaydogdu — CHANGES_REQUESTED** [review]

**Review by moaydogdu — APPROVED** [review]
LGTM!

**Review by ibrahimdenizz — CHANGES_REQUESTED** [review]

File: src/test/java/com/ays/assignment/service/impl/AssignmentSearchServiceImplTest.java [hunk 2]
```diff
@@ -184,4 +185,45 @@ void givenInValidAssignmentSearchRequest_whenAssignmentNotExistForSearching_then
...
+                AysAssignmentUserAlreadyAssigned.class,
+                () -> assignmentSearchService.searchAssignment(mockAssignmentSearchRequest)
+        );
+
+        Mockito.verify(identity, Mockito.times(1)).getUserId();
+        Mockito.verify(identity, Mockito.times(1)).getInstitutionId();
+        Mockito.verify(userRepository, Mockito.times(1))
+                .findByIdAndInstitutionId(mockUserEntity.getId(), mockUserEntity.getInstitutionId());
+        Mockito.verify(assignmentRepository, Mockito.never())
+                .findNearestAvailableAssignment(Mockito.any(Point.class), Mockito.anyString());
```
[the remark this answers is not in the document]
Comment (AhmetAksunger) [reply src/test/java/com/ays/assignment/service/impl/AssignmentSearchServiceImplTest.java:233]: Yeni committe çözüldü.

File: src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.ays.assignment.util.exception;
+
+import com.ays.common.util.exception.AysAlreadyException;
+
+import java.io.Serial;
+
```
[the remark this answers is not in the document]
Comment (AhmetAksunger) [reply src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java:6]: Yeni committe çözüldü.

File: src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+public class AysAssignmentUserAlreadyAssigned extends AysAlreadyException {
+
+    @Serial
+    private static final long serialVersionUID = -3240549605152804148L;
```
[the remark this answers is not in the document]
Comment (AhmetAksunger) [reply src/main/java/com/ays/assignment/util/exception/AysAssignmentUserAlreadyAssigned.java:16]: Yeni committe çözüldü.

**Review by ibrahimdenizz — DISMISSED** [review]

File: src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java [hunk 5]
```diff
@@ -40,6 +41,10 @@ public Assignment searchAssignment(AssignmentSearchRequest searchRequest) {
                 .findByIdAndInstitutionId(userId, institutionId)
                 .orElseThrow(() -> new AysUserNotExistByIdException(userId));
 
+        if (assignmentRepository.existsByUserId(userId)) {
+            throw new AysAssignmentUserAlreadyAssigned(userId);
+        }
```
Comment (agitrubard) [inline src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java:46]: Burada assignment'ın statü kontrolü de yapılmalı, eğer statüsü `COMPLETED` değil ise buradaki hata fırlatılmalı. Issue içerisine bu detayı yazmamışım, issue içerisini de güncelledim ✅

**Review by agitrubard — CHANGES_REQUESTED** [review]

File: src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java (hunk 5 above)
Comment (AhmetAksunger) [reply src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java:47]: Hallettim.

**Review by ibrahimdenizz — DISMISSED** [review]

File: src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java (hunk 5 above)
Comment (agitrubard) [reply src/main/java/com/ays/assignment/service/impl/AssignmentSearchServiceImpl.java:47]: Buna birlikte bakalım @AhmetAksunger @anilhazar

**Review by agitrubard — CHANGES_REQUESTED** [review]

**Review by agitrubard — APPROVED** [review]

**Review by ibrahimdenizz — APPROVED** [review]

**Pull request merged**